runner/providers/v0.1.1/external.go

```diff
@@ -0,0 +1,387 @@
+package v011
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"os/exec"
+
+	"github.com/pkg/errors"
+
+	garmErrors "github.com/cloudbase/garm-provider-common/errors"
+	execution "github.com/cloudbase/garm-provider-common/execution/v0.1.1"
+	commonParams "github.com/cloudbase/garm-provider-common/params"
+	garmExec "github.com/cloudbase/garm-provider-common/util/exec"
+	"github.com/cloudbase/garm/config"
+	"github.com/cloudbase/garm/metrics"
+	"github.com/cloudbase/garm/params"
+	"github.com/cloudbase/garm/runner/common"
+)
+
+var _ common.Provider = (*external)(nil)
+
+func NewProvider(ctx context.Context, cfg *config.Provider, controllerID string) (common.Provider, error) {
+	if cfg.ProviderType != params.ExternalProvider {
+		return nil, garmErrors.NewBadRequestError("invalid provider config")
+	}
+
+	execPath, err := cfg.External.ExecutablePath()
+	if err != nil {
+		return nil, errors.Wrap(err, "fetching executable path")
+	}
+
+	// Set GARM_INTERFACE_VERSION to the version of the interface that the external
+	// provider implements. This is used to ensure compatibility between the external
+	// provider and garm
+	envVars := cfg.External.GetEnvironmentVariables()
+	envVars = append(envVars, fmt.Sprintf("GARM_INTERFACE_VERSION=%s", cfg.External.InterfaceVersion))
+
+	return &external{
+		ctx:                  ctx,
+		controllerID:         controllerID,
+		cfg:                  cfg,
+		execPath:             execPath,
+		environmentVariables: envVars,
+	}, nil
+}
+
+type external struct {
+	ctx                  context.Context
+	cfg                  *config.Provider
+	controllerID         string
+	execPath             string
+	environmentVariables []string
+}
+
+func (e *external) validateResult(inst commonParams.ProviderInstance) error {
+	if inst.ProviderID == "" {
+		return garmErrors.NewProviderError("missing provider ID")
+	}
+
+	if inst.Name == "" {
+		return garmErrors.NewProviderError("missing instance name")
+	}
+
+	if !IsValidProviderStatus(inst.Status) {
+		return garmErrors.NewProviderError("invalid status returned (%s)", inst.Status)
+	}
+
+	return nil
+}
+
+// CreateInstance creates a new compute instance in the provider.
+func (e *external) CreateInstance(ctx context.Context, bootstrapParams commonParams.BootstrapInstance, createInstanceParams common.CreateInstanceParams) (commonParams.ProviderInstance, error) {
+	extraspecs := createInstanceParams.CreateInstanceV011.PoolInfo.ExtraSpecs
+	extraspecsValue, err := json.Marshal(extraspecs)
+	if err != nil {
+		return commonParams.ProviderInstance{}, errors.Wrap(err, "serializing extraspecs")
+	}
+	asEnv := []string{
+		fmt.Sprintf("GARM_COMMAND=%s", execution.CreateInstanceCommand),
+		fmt.Sprintf("GARM_CONTROLLER_ID=%s", e.controllerID),
+		fmt.Sprintf("GARM_POOL_ID=%s", bootstrapParams.PoolID),
+		fmt.Sprintf("GARM_PROVIDER_CONFIG_FILE=%s", e.cfg.External.ConfigFile),
+		fmt.Sprintf("GARM_POOL_IMAGE=%s", createInstanceParams.CreateInstanceV011.PoolInfo.Image),
```

Comment:
I think this is already set in the bootstrap we send to stdin.

Reply:
Indeed, removed it for CreateInstance. For the moment, changed it to `CreateInstance(ctx context.Context, bootstrapParams commonParams.BootstrapInstance, _ common.CreateInstanceParams)` since we don't use (for now) anything from there